src/ast_calc.c

```diff
@@ -1,46 +1,216 @@
-#include "ast_calc.h"
-#include "ast.h"
+#include <stdio.h>
+#include <stdlib.h>
+#include "abstract_calc.h"
+#include "config.h"
+
+typedef enum {
+    NT_NUM,
+    NT_PLUS,
+    NT_MINUS,
+    NT_MUL,
+    NT_DIV,
+    NT_NEG,
+} node_type_t;
+
+typedef struct ast_node_t {
+    union {
+        calc_value_t value;
+        struct {
+            int left;
+            int right;
+        };
+    };
+    node_type_t node_type;
+} ast_node_t;
+
+typedef struct arena_node_t {
+    ast_node_t *ast;
+    size_t capacity;
+    size_t allocated;
+} arena_node_t;
+
+#define ARENA_INIT_CAPACITY 16
+#define ARENA_CAPACITY_GROWTH 2
+
+static int
+arena_init (arena_node_t *arena)
+{
+    arena->capacity = ARENA_INIT_CAPACITY;
+    arena->allocated = 0;
+    arena->ast = malloc (arena->capacity * sizeof (*arena->ast));
+    if (NULL == arena->ast) {
+        perror ("error: cannot initialize arena");
+        return (EXIT_FAILURE);
+    }
+
+    return (EXIT_SUCCESS);
+}
 
-#define EVAL(value) YYSTYPE *res = yyget_extra (scanner); *res = value;
-#define EVAL_NUM(node)        node_value_create (*((calc_value_t *) &node));
-#define EVAL_ADD(left, right) node_op_create (NT_PLUS, left, right)
-#define EVAL_SUB(left, right) node_op_create (NT_MINUS, left, right)
-#define EVAL_MUL(left, right) node_op_create (NT_MUL, left, right)
-#define EVAL_DIV(left, right) node_op_create (NT_DIV, left, right)
-#define EVAL_NEG(value)       node_op_create (NT_NEG, value, NULL)
+static void
+arena_destroy (arena_node_t *arena)
+{
+    if (NULL != arena->ast) {
+        free (arena->ast);
+    }
+
+    arena->ast = NULL;
+}
+
+static int
+arena_allocate (arena_node_t *arena, size_t alloc_size)
+{
+    size_t new_alloc_size = arena->allocated + alloc_size;
+
+    if (new_alloc_size > arena->capacity) {
+        size_t new_capacity = ARENA_CAPACITY_GROWTH * arena->capacity;
```

Comment:
Potentially new_alloc_size might be greater than new_capacity. It make sense to use new_allocs_size as a base size.